txnsync/outgoing_test.go

```diff
@@ -0,0 +1,294 @@
+	logging.Logger
+}
+
+func (m mockAsyncLogger) outgoingMessage(mstat msgStats) {
+}
+
+func (m mockAsyncLogger) incomingMessage(mstat msgStats) {
+}
+
+func (m mockAsyncLogger) Infof(string, ...interface{}) {}
+
+func TestAsyncMessageSent(t *testing.T) {
+	a := require.New(t)
+
+	var s syncState
+	s.clock = timers.MakeMonotonicClock(time.Now())
+	s.log = mockAsyncLogger{}
+
+	asyncEncoder := messageAsyncEncoder{
+		state: &s,
+		messageData: sentMessageMetadata{
+			encodedMessageSize:  0,
+			sentTranscationsIDs: []transactions.Txid{},
+			message: &transactionBlockMessage{
+				Version: txnBlockMessageVersion,
+				Round:   1,
+			},
+			peer:              &Peer{},
+			sentTimestamp:     0 * time.Millisecond,
+			sequenceNumber:    0,
+			partialMessage:    false,
+			filter:            bloomFilter{},
+			transactionGroups: []transactions.SignedTxGroup{},
+		},
+		roundClock:           timers.MakeMonotonicClock(time.Now()),
+		peerDataExchangeRate: 0,
+	}
+
+	oldTimestamp := asyncEncoder.messageData.sentTimestamp
+	a.Equal(asyncEncoder.asyncMessageSent(false, 0), errTransactionSyncOutgoingMessageSendFailed)
+	err := asyncEncoder.asyncMessageSent(true, 1337)
+	a.Equal(err, errTransactionSyncOutgoingMessageQueueFull)
+	a.NotEqual(asyncEncoder.messageData.sentTimestamp, oldTimestamp)
+	a.Equal(asyncEncoder.messageData.sequenceNumber, uint64(1337))
+
+	// Make this buffered for now so we catch the select statement
+	asyncEncoder.state.outgoingMessagesCallbackCh = make(chan sentMessageMetadata, 1)
+
+	err = asyncEncoder.asyncMessageSent(true, 1337)
+	a.Nil(err)
+}
+
+type mockAsyncNodeConnector struct {
+	NodeConnector
+	called *bool
+}
+
+func (m mockAsyncNodeConnector) SendPeerMessage(netPeer interface{}, msg []byte, callback SendMessageCallback) {
+	*m.called = true
+}
+
+func (m mockAsyncNodeConnector) GetPendingTransactionGroups() (txGroups []transactions.SignedTxGroup, latestLocallyOriginatedGroupCounter uint64) {
+	return []transactions.SignedTxGroup{}, 1
+}
+
+func TestAsyncEncodeAndSend(t *testing.T) {
+	a := require.New(t)
+
+	var s syncState
+	s.clock = timers.MakeMonotonicClock(time.Now())
+	s.log = mockAsyncLogger{}
+	called := false
+	s.node = mockAsyncNodeConnector{called: &called}
+	s.messageSendWaitGroup = sync.WaitGroup{}
+
+	asyncEncoder := messageAsyncEncoder{
+		state: &s,
+		messageData: sentMessageMetadata{
+			encodedMessageSize:  0,
+			sentTranscationsIDs: []transactions.Txid{},
+			message: &transactionBlockMessage{
+				Version: txnBlockMessageVersion,
+				Round:   1,
+			},
+			peer:              &Peer{},
+			sentTimestamp:     0 * time.Millisecond,
+			sequenceNumber:    0,
+			partialMessage:    false,
+			filter:            bloomFilter{},
+			transactionGroups: []transactions.SignedTxGroup{},
+		},
+		roundClock:           timers.MakeMonotonicClock(time.Now()),
+		peerDataExchangeRate: 0,
+	}
+
+	asyncEncoder.state.messageSendWaitGroup.Add(1)
+
+	err := asyncEncoder.asyncEncodeAndSend(nil)
```

Comment:
asyncEncodeAndSend has 2 more cases which are not tested: 
- len(encoder.messageData.transactionGroups) > 0 
- encodeTransactionGroups returns error